src/Service.Tests/Unittests/MultipleCreateOrderHelperUnitTests.cs

```diff
@@ -0,0 +1,367 @@
+                targetEntityName: targetEntityName,
+                metadataProvider: _sqlMetadataProvider,
+                columnDataInSourceBody: columnDataInSourceBody,
+                targetNodeValue: targetNodeValue,
+                nestingLevel: 1));
+        }
+
+        /// <summary>
+        /// Test to validate that when all the relationship fields in the source/target entity are non-autogenerated,
+        /// either the input for the source or the target entity should contain the value for the relationship fields.
+        /// The entity which contains the values for all relationship fields is considered as the referenced entity, and the other
+        /// entity is considered as the referencing entity.
+        ///
+        /// Here, the relationship between entities: 'User_NonAutogenRelationshipColumn' and 'UserProfile_NonAutogenRelationshipColumn'
+        /// is defined as User_NonAutogenRelationshipColumn(username) -> UserProfile_NonAutogenRelationshipColumn(username)
+        /// where the field 'username' is non-autogenerated in both the entities.
+        /// </summary>
+        [TestMethod]
+        public void ValidateDeterministicReferencingEntityForNonAutogenRelationshipColumns()
+        {
+            // Test 1: The value for relationship field 'username' is present in the input for the source entity.
+            IMiddlewareContext context = SetupMiddlewareContext();
+            string sourceEntityName = "User_NonAutogenRelationshipColumn";
+            string targetEntityName = "UserProfile";
+
+            // Setup column input in source entity.
+            Dictionary<string, IValueNode> columnDataInSourceBody = new()
+            {
+                { "username", new StringValueNode("DAB") },
+                { "email", new StringValueNode("dab@microsoft.com") }
+            };
+
+            // Setup column input in source entity.
+            ObjectValueNode targetNodeValue = new();
+            List<ObjectFieldNode> fields = new()
+            {
+                new ObjectFieldNode("profilepictureurl", "dab/profilepicture"),
+                new ObjectFieldNode("userid", 10)
+            };
+            targetNodeValue = targetNodeValue.WithFields(fields);
+
+            // Get the referencing entity name. Since the source entity contained the value for relationship field,
+            // it act as the referenced entity, and the target entity act as the referencing entity.
+            string referencingEntityName = MultipleCreateOrderHelper.GetReferencingEntityName(
+                relationshipName: "UserProfile_NonAutogenRelationshipColumn",
+                context: context,
+                sourceEntityName: sourceEntityName,
+                targetEntityName: targetEntityName,
+                metadataProvider: _sqlMetadataProvider,
+                columnDataInSourceBody: columnDataInSourceBody,
+                targetNodeValue: targetNodeValue,
+                nestingLevel: 1);
```

Comment:
Added a comment to the test..

`nestingLevel` is included in the error messages to provide useful information so that users can easily identify where the problem in their mutation request is. It does not have an effect in the order determination (or any other) logic.